src/SupabaseClient.ts

```diff
@@ -0,0 +1,191 @@
+
+const DEFAULT_OPTIONS = {
+  schema: 'public',
+  autoRefreshToken: true,
+  persistSession: true,
+  detectSessionInUrl: true,
+  headers: DEFAULT_HEADERS,
+}
+
+/**
+ * Supabase Client.
+ *
+ * An isomorphic Javascript client for interacting with Postgres.
+ */
+export default class SupabaseClient {
+  schema: string
+  restUrl: string
+  realtimeUrl: string
+  authUrl: string
+  auth: GoTrueClient
+  realtime: RealtimeClient
+
+  /**
+   * Create a new client for use in the browser.
+   * @param supabaseUrl The unique Supabase URL which is supplied when you create a new project in your project dashboard.
+   * @param supabaseKey The unique Supabase Key which is supplied when you create a new project in your project dashboard.
+   * @param options.schema You can switch in between schemas. The schema needs to be on the list of exposed schemas inside Supabase.
+   * @param options.autoRefreshToken Set to "true" if you want to automatically refresh the token before expiring.
+   * @param options.persistSession Set to "true" if you want to automatically save the user session into local storage.
+   * @param options.detectSessionInUrl Set to "true" if you want to automatically detects OAuth grants in the URL and signs in the user.
+   * @param options.headers Any additional headers to send with each network request.
+   */
+  constructor(
+    public supabaseUrl: string,
+    public supabaseKey: string,
+    options?: SupabaseClientOptions
+  ) {
+    if (!supabaseUrl) throw new Error('supabaseUrl is required.')
+    if (!supabaseKey) throw new Error('supabaseKey is required.')
+
+    const settings = { ...DEFAULT_OPTIONS, ...options }
+    this.restUrl = `${supabaseUrl}/rest/v1`
+    this.realtimeUrl = `${supabaseUrl}/realtime/v1`.replace('http', 'ws')
+    this.authUrl = `${supabaseUrl}/auth/v1`
+    this.schema = settings.schema
+
+    this.auth = this._initGoTrueClient(settings)
+    this.realtime = this._initRealtimeClient()
+
+    this.realtime.onOpen(() => console.log('OPEN'))
+    this.realtime.onClose(() => console.log('CLOSED'))
+    this.realtime.onError((e: Error) => console.log('Socket error', e))
+  }
+
+  /**
+   * Perform a table operation.
+   *
+   * @param table The table name to operate on.
+   */
+  from<T = any>(tableName: string) {
+
+    // At this point, we don't know whether the user is going to
+    // make a call to Realtime or to PostgREST, so we need to do
+    // an intermdiary step where we return both.
+    // We have to make sure "this" is bound correctly on each part.
+    // let rest = this._initPostgRESTClient()
+
+    // let rest = new SupaabseQueryBuilder(this.restUrl || '', {})
+    let rest = this._initPostgRESTClient()
+    let subscription = new SupabaseRealtimeClient(this.realtime, this.schema, tableName)
+
+    const builder: SupabaseQueryClient<T> = {
+      rest,
+      subscription,
+      select: (columns) => {
+        return rest.from(tableName).select(columns)
```

Comment:
this needs to be `return rest.from<T>(tableName).select(columns)` to make sure we pass the type down to postgrest-js. 

Did you have a chance to look into the `SupabaseQueryClient<T> extends PostgrestQueryBuilder<T>` route instead? If that doesn't work this is probably fine as long as we pass down the `<T>` in all the `from`s.

Reply:
This reminds me, I think we wanted to call it `SupabaseQueryBuilder` instead of `SupabaseQueryClient`, no?

Reply:
Thanks for this - still getting my head around passing through the generics.

Also, do you know why the DocComments don't show up in the intellisense once we get down to the PostgrestFilterBuilder?

![image](https://user-images.githubusercontent.com/10214025/96529859-c860c680-12b8-11eb-8eb2-af10f7f1913c.png)

Reply:
I've pushed the latest changes, and adding tests now, but for the life of me I can't figure out why the DocComments aren't getting picked up. @soedirgo do you know why?

FYI, to make it easier, I added a ToDo app in the `example` folder . To use it:

- Run `npm run build`
- Create an app in Supabase and run the To Do List quickstart
- Go into /example 
  - add a `.env.local` file and add `NEXT_PUBLIC_SUPABASE_URL` and `NEXT_PUBLIC_SUPABASE_KEY`
  -`npm run dev`

This example covers all the functionality, but I can't get the Intellisense to come through 😠

Reply:
> I've pushed the latest changes, and adding tests now, but for the life of me I can't figure out why the DocComments aren't getting picked up. 

It does for me
![image](https://user-images.githubusercontent.com/5748289/96587286-bad92a00-1314-11eb-8a80-96c8585fbef1.png)
You might have to hit `control + space` in order for vscode to fully expand them.

Reply:
> You might have to hit control + space in order for vscode to fully expand them.

Wow. you should tweet this. TIL

Reply:
Ah, looks like it was resolved. `select()` was giving me `any` earlier so I was missing autocompletion.

FWIW, I managed to do this by extending `PostgrestQueryBuilder<T>`:

```ts
// SupabaseClient.ts
const builder = rest.from<T>(tableName) as SupabaseQueryBuilder<T>
builder.subscription = subscription
builder.realtime = this.realtime

// types.ts
export class SupabaseQueryBuilder<T> extends PostgrestQueryBuilder<T> {
  subscription!: SupabaseRealtimeClient
  realtime!: RealtimeClient

  on(event: 'INSERT' | 'UPDATE' | 'DELETE' | '*', callback: Function) {
    if (!this.realtime.isConnected()) {
      this.realtime.connect()
    }
    return this.subscription.on(event, callback)
  }
}

```

Reply:
>  `subscription!`

Damn there's no end to obscure TypeScript syntax. This seems better Bobbie, thanks!